lib/vfscore: Use `uk_syscall_do` for syscall delegation

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Since `uk_syscall_r_` symbols tend to also invoke the system call enter
and exit tables if the `syscall_shim` library is enabled, replace such 
calls with the `uk_syscall_do_` symbol variant which does not involve  
any tables.

mogasergiu committed:
lib/vfscore: Use `uk_syscall_do_`  instead of `uk_syscall_r_` symbols

Since `uk_syscall_r_` symbols tend to also invoke the system call enter
and exit tables if the `syscall_shim` library is enabled, replace such
calls with the `uk_syscall_do_` symbol variant which does not involve
any tables.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented:
NOTE: Hastily submitted so we can merge #1277. Thorough review required.

mogasergiu commented:
Rebase fpush

andreittr reviewed (changes requested):
GH won't let me comment on a line not in the diff :roll_eyes:.
I think you also need to patch `uk_syscall_e_open` in the `creat` syscall impl @ line 241.

mogasergiu commented:
> GH won't let me comment on a line not in the diff 🙄. I think you also need to patch `uk_syscall_e_open` in the `creat` syscall impl @ line 241.

Good catch, I only just `%s/uk_syscall_r_/uk_syscall_do_`. I forgot about the wrongly registered system calls (not using the raw variant).

However, actually, I am not so sure. This  is could not the case here. Note that `creat` is **sadly** registered with the `UK_SYSCALL_DEFINE` variant which handles `errno` differently. However, the `open` syscall is registered with `UK_LLSYSCALL_R_DEFINE`, which handles it the exact opposite way. Correct me if I am wrong though. I am afraid that we either have to convert `creat` to `UK_SYSCALL_R_DEFINE` or simply accept that it will end up calling into the tables twice.

andreittr commented:
> Note that creat is sadly registered with the `UK_SYSCALL_DEFINE` variant which handles errno differently

Oooof, I had not noticed that; how many syscall definition macros do we have? :smile: 
Considering every other similar syscall in vfscore is `_R`, I think converting to `UK_SYSCALL_R_DEFINE` + using `_do_open` is the best course here.

mogasergiu committed:
lib/vfscore: Register `creat` using the raw variant

The `creat` has been registered as non-raw a long time ago, back
when the registration policy was different. Now, all system calls
should be registered as raw. Do so for `creat` as well and make it
call the `uk_syscall_do_` variant of `open` in order to avoid
invoking the system call enter/exit tables twice.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

andreittr reviewed (approved):
Alright, looks good, thanks!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Approved-by: Andrei Tatar <andrei@unikraft.io>
